File: windows/windows-build/packaging/map_language.js
/* Local UI language only. Fixed application labels are separate from user content. */
'use strict';
(() => {
  const fs=require('fs'),path=require('path');
  const folder=process.env.RADIOASSISTENT_DATA||path.join(process.env.APPDATA,'Radioassistent');
  const file=path.join(folder,'ui-language.json');
  let code=process.env.HAMNAVIGATOR_LANGUAGE;
  if(!['en','nb','sv'].includes(code))try{code=JSON.parse(fs.readFileSync(file,'utf8')).language;}catch{}
  window.hnLanguage=code==='sv'?'sv':code==='en'?'en':'nb';window.hnEnglish=window.hnLanguage==='en';
  const labels={
    'Min stasjon':'My station','⌖ Min stasjon':'⌖ My station','Projeksjon':'Projection','Kallesignaler':'Callsigns','Fest punkt':'Place pin','Rutenett':'Grid','Statistikk':'Statistics','Grålinje':'Grey line','Månen':'Moon','Snarveier':'Shortcuts','Varsellyd':'Alert sound','Oppsett':'Settings','Importer logg':'Import log','Les lokal logg':'Read local log','Tidssoner':'Time zones','Lokatorlag':'Grid overlay','Spottere':'Spotters','Åpne':'Open','‹ Tilbake':'‹ Back',
    'Stasjon og tilkobling':'Station and connection','Arbeidsflyt':'Workflow','Kallesignaloppslag':'Callsign lookup','Lyd og tale':'Audio and speech','Kartvisning':'Map display','Lokatorruter':'Grid squares','Loggintegrasjoner':'Log integrations','Lydvarsler':'Audio alerts','Egne varsler':'Custom alerts','Loggbok':'Logbook','Om HamNavigator':'About HamNavigator',
    'Bruksanvisning':'User guide','HamNavigator Map – bruksanvisning':'HamNavigator Map – user guide','Lukk':'Close','Søkbar bruksanvisning for HamNavigator Map':'Searchable user guide for HamNavigator Map',
    'Åpne Map fra HamNavigator for felles leveringsstatus.':'Open Map from HamNavigator for shared delivery status.',
    'Levering kunne ikke legges i kø. Kontakt lagret lokalt. Se Status og sikkerhetskopi.':'Could not queue delivery. The contact is saved locally. See Status and backup.',
    'Logglevering registrert i HamNavigator. Se Status og sikkerhetskopi.':'Log delivery registered in HamNavigator. See Status and backup.'
  };
  const swedish=/* HN_SWEDISH_CATALOG */{};
  window.hnText=source=>window.hnLanguage==='sv'?(swedish[labels[source]||source]||source):window.hnEnglish?(labels[source]||source):source;
  document.addEventListener('DOMContentLoaded',()=>{
    const header=document.querySelector('#hn-map-header');if(!header)return;
    const nav=header.querySelector('nav');
    if(window.hnLanguage!=='nb'){
      const subtitle=header.querySelector('.hn-map-brand span');if(subtitle)subtitle.textContent=window.hnLanguage==='sv'?'KARTA / STATIONSÖVERSIKT':'MAP / STATION OVERVIEW';
      nav.setAttribute('aria-label',window.hnLanguage==='sv'?'Huvudverktyg':'Main tools');
      for(const button of nav.querySelectorAll('button'))button.textContent=window.hnText(button.textContent);
    }
    const button=document.createElement('button');button.textContent='Språk / Language';button.id='hn-language-button';nav.append(button);
    button.onclick=()=>{
      const dialog=document.createElement('dialog');dialog.setAttribute('aria-label','Språk / Language');
      Object.assign(dialog.style,{background:'#102332',color:'#eef5ff',border:'1px solid #6c96ad',borderRadius:'10px',maxWidth:'500px',padding:'24px'});
      const title=document.createElement('h2');title.textContent='Språk / Language';
      const select=document.createElement('select');select.setAttribute('aria-label','Språk / Language');
      for(const [value,label] of [['nb','Norsk (bokmål)'],['en','English'],['sv','Svenska']]){const option=document.createElement('option');option.value=value;option.textContent=label;select.append(option);}
      select.value=window.hnLanguage;
      const note=document.createElement('p');note.textContent=window.hnLanguage==='sv'?'Språket gäller HamNavigator, MY SHACK och Map. Spara, stoppa sändningen och starta om HN.':window.hnEnglish?'The language applies to HamNavigator, MY SHACK and Map. Save, stop transmission and restart HN.':'Språket gjelder HamNavigator, MY SHACK og Map. Lagre, stopp sending og start HN på nytt.';
      const status=document.createElement('p');status.setAttribute('role','status');
      const save=document.createElement('button');save.textContent=window.hnLanguage==='sv'?'Spara språk':window.hnEnglish?'Save language':'Lagre språk';
      const close=document.createElement('button');close.textContent=window.hnText('Lukk');close.onclick=()=>dialog.close();
      save.onclick=()=>{
        const temporary=file+'.'+require('crypto').randomBytes(6).toString('hex')+'.tmp';
        try{
          fs.mkdirSync(folder,{recursive:true});fs.writeFileSync(temporary,JSON.stringify({language:select.value}),{encoding:'utf8',flag:'wx'});fs.renameSync(temporary,file);
          status.textContent=select.value==='sv'?'Sparat. Stoppa sändningen och starta om HN.':select.value==='en'?'Saved. Stop transmission and restart HN.':'Lagret. Stopp sending og start HN på nytt.';
        }catch{status.textContent=window.hnLanguage==='sv'?'Språket kunde inte sparas. Försök igen i MY SHACK.':window.hnEnglish?'Could not save the language. Try again in MY SHACK.':'Språket kunne ikke lagres. Prøv igjen i MY SHACK.';}
        finally{try{fs.unlinkSync(temporary);}catch{}}
      };
      dialog.append(title,select,note,save,close,status);document.body.append(dialog);
      dialog.addEventListener('close',()=>{dialog.remove();button.focus();},{once:true});dialog.showModal();select.focus();
    };
  });
})();

File: windows/windows-build/packaging/map_language.py
"""Install language controls while retaining native Map translations and stable handlers."""
from pathlib import Path
import shutil,json
HERE=Path(__file__).resolve().parent
def install(bundle):
    renderer=Path(bundle)/'release/HamNavigator-Map/resources/app/src/renderer'
    script=(HERE/'map_language.js').read_text(encoding='utf-8')
    dictionary=json.loads((HERE/'map-language-sv.json').read_text(encoding='utf-8'))
    assert '/* HN_SWEDISH_CATALOG */{}' in script
    (renderer/'lib/hamnavigator-language.js').write_text(script.replace('/* HN_SWEDISH_CATALOG */{}',json.dumps(dictionary,ensure_ascii=False)),encoding='utf-8')
    shutil.copy2(HERE/'map-i18n-sv.json',renderer.parents[1]/'resources/i18n/sv.json')
    p=renderer/'lib/gt.js';s=p.read_text(encoding='utf-8')
    if 'sv: "i18n/sv.json"' not in s:s=s.replace('GT.languages = {','GT.languages = {\n  sv: "i18n/sv.json",',1)
    p.write_text(s,encoding='utf-8')
    for p in [renderer/'GridTracker2.html',*renderer.glob('gt_*.html')]:
        s=p.read_text(encoding='utf-8')
        if 'lib/hamnavigator-language.js' not in s:
            assert '<head>' in s
            p.write_text(s.replace('<head>','<head>\n<script src="./lib/hamnavigator-language.js"></script>',1),encoding='utf-8')
    p=renderer/'lib/i18n.js';s=p.read_text(encoding='utf-8')
    if 'if(window.hnEnglish)locale=' not in s and 'if(["en","sv"].includes(window.hnLanguage))' not in s:
        s=s.replace('function readLocaleFile(locale)\n{','function readLocaleFile(locale)\n{\n  if(window.hnEnglish)locale="en";')
        p.write_text(s,encoding='utf-8')
    s=p.read_text(encoding='utf-8').replace('if(window.hnEnglish)locale="en";','if(["en","sv"].includes(window.hnLanguage))locale=window.hnLanguage;')
    p.write_text(s,encoding='utf-8')
    for name in ('hamnavigator-ui.js','hamnavigator-menus.js'):
        p=renderer/'lib'/name;s=p.read_text(encoding='utf-8')
        s=s.replace('text.textContent=label;','text.textContent=window.hnText(label);')
        s=s.replace("b.textContent='‹ Tilbake';","b.textContent=window.hnText('‹ Tilbake');")
        s=s.replace('button.textContent=labels[target[1]];','button.textContent=window.hnText(labels[target[1]]);')
        p.write_text(s,encoding='utf-8')
    p=renderer/'lib/hamnavigator-manual.js';s=p.read_text(encoding='utf-8')
    for value in ('Bruksanvisning','HamNavigator Map – bruksanvisning','Lukk','Søkbar bruksanvisning for HamNavigator Map'):
        s=s.replace("='"+value+"';","=window.hnText('"+value+"');")
        s=s.replace("setAttribute('aria-label','"+value+"')","setAttribute('aria-label',window.hnText('"+value+"'))")
    s=s.replace("frame.src='./HamNavigator-Map-bruksanvisning.html';","frame.src=window.hnEnglish?'./HamNavigator-Map-bruksanvisning.en.html':'./HamNavigator-Map-bruksanvisning.html';")
    s=s.replace("window.hnEnglish?'./HamNavigator-Map-bruksanvisning.en.html':'./HamNavigator-Map-bruksanvisning.html'","'./HamNavigator-Map-bruksanvisning'+(window.hnLanguage==='nb'?'':'.'+window.hnLanguage)+'.html'")
    p.write_text(s,encoding='utf-8')
    p=renderer/'lib/hamnavigator-delivery.js';s=p.read_text(encoding='utf-8')
    for text,colour in [('Åpne Map fra HamNavigator for felles leveringsstatus.','orange'),('Levering kunne ikke legges i kø. Kontakt lagret lokalt. Se Status og sikkerhetskopi.','orange'),('Logglevering registrert i HamNavigator. Se Status og sikkerhetskopi.','white')]:
        old="'<font style=\"color:"+colour+'">'+text+"</font>'"
        new=json.dumps('<font style="color:'+colour+'">')+'+window.hnText('+json.dumps(text,ensure_ascii=False)+')+"</font>"'
        s=s.replace(old,new)
    p.write_text(s,encoding='utf-8')
    print('Map language selector and Norwegian, English, Swedish menus installed')
if __name__=='__main__':install(HERE.parents[1]/'bygg/klient')
